File: DSA/Linked_Lists/singly/javascript/linkedList.js
#!/usr/bin/env node

/**
 * singly_linked_list.js
 *
 * This module defines a basic singly linked list data structure in JavaScript.
 */

class Node {
  /**
   *
   * @param {*} val : Data to be Hold into the Node.
   */
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /**
   *
   * @param {*} val : Data to be pushed within the New Node.
   * @returns {LinkedList} : The whole linked list after adding the new Node.
   */
  push(val) {
    let newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }

    this.length++;
    return this;
  }

  /**
   *
   * @returns : The Popped Node.
   */
  pop() {
    if (!this.head) return undefined;

    let currentNode = this.head;
    let newTail = currentNode;
    while (currentNode.next) {
      newTail = currentNode;
      currentNode = currentNode.next;
    }
    newTail.next = null;
    this.tail = newTail;

    this.length--;

    if (this.length === 0) {
      this.head = null;
      this.tail = null;
    }

    return currentNode;
  }

  /**
   *
   * @returns : The Node that will be shifted (removed at beginning)
   */
  shift() {
    if (!this.head) return undefined;

    let headNode = this.head;
    this.head = headNode.next;
    this.length--;

    if (this.length === 0) {
      this.tail = null;
    }

    return headNode;
  }

  /**
   *
   * @param {*} val : The New Node data to be Add on the first.
   * @returns : The whole linked List.
   */
  unShift(val) {
    const newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head = newNode;
    }

    this.length++;
    return this;
  }

  /**
   *
   * @param {Number} index : Index to indicate which Element should be returned.
   * @returns : the corresponding Node.
   */
  get(index) {
    if (index < 0 || index >= this.length) return null;
    let currentNode = this.head;
    let counter = 0;
    while (counter < index) {
      currentNode = currentNode.next;
      counter++;
    }

    return currentNode;
  }

  /**
   *
   * @param {Number} index : Index to indicate which Element should be Modified.
   * @param {any} data : The Data which will be the new Value of the Node.
   * @returns : the corresponding Node.
   */
  set(index, data) {
    const node = this.get(index);
    if (!node) return null;
    node.val = data;
    return node;
  }

  /**
   *
   * @param {Number} index : The index to insert a new node.
   * @param {*} val : The Data to be inserted within the node.
   * @returns : The whole linked list.
   */
  insert(index, val) {
    if (index < 0 || index > this.length) return null;

    if (index === 0) {
      this.unShift(val);
      return this;
    }

    if (index === this.length) {
      this.push(val);
      return this;
    }

    const newNode = new Node(val);
    const prevNode = this.get(index - 1);

    newNode.next = prevNode.next;
    prevNode.next = newNode;

    this.length++;
    return this;
  }

  /**
   *
   * @param {number} index - The index of the node to remove.
   * @returns {Node|null} - The removed node, or null if index is invalid.
   */
  remove(index) {
    if (index < 0 || index >= this.length) return null;

    // Remove head
    if (index === 0) return this.shift();

    // Remove tail
    if (index === this.length - 1) return this.pop();

    // Remove in the middle
    const prevNode = this.get(index - 1);
    const removedNode = prevNode.next;

    prevNode.next = removedNode.next;
    this.length--;

    return removedNode;
  }

  /**
   *
   * @returns : The whole list values into an Array type.
   */
  toArray() {
    const result = [];
    let current = this.head;
    while (current) {
      result.push(current.val);
      current = current.next;
    }

    return result;
  }

  /**
   *
   * @returns : the whole list as an array type printed.
   */
  print() {
    return console.log(this.toArray);
  }

  /**
   *
   * @returns : The whole linked list reversed.
   */
  reverse() {
    if (!this.head || !this.head.next) return this;

    let current = this.head;
    let prev = null;
    let next = null;

    this.head = this.tail;

    while (current) {
      next = current.next;
      current.next = prev;
      prev = current;
      current = next;
    }

    this.head = prev;

    return this;
  }
}

File: DSA/Linked_Lists/singly/python/linked_list.py
#!/usr/bin/env python3

"""
singly_linked_list.py

This module defines a basic singly linked list data structure in Python.
"""

from typing import Any, Optional


class Node:
    """
    A node in a singly linked list.

    Attributes:
        val: The data stored in the node.
        next: The reference to the next node.
    """

    def __init__(self, val: Any) -> None:
        self.val: Any = val
        self.next: Optional['Node'] = None


class LinkedList:
    """
    Singly Linked List implementation.
    """

    def __init__(self) -> None:
        self.head: Optional[Node] = None
        self.tail: Optional[Node] = None
        self.length: int = 0

    def push(self, val: Any) -> 'LinkedList':
        """
        Adds a new node with the given value to the end of the list.
        """
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            assert self.tail is not None
            self.tail.next = new_node
            self.tail = new_node

        self.length += 1
        return self

    def pop(self) -> Optional[Node]:
        """
        Removes the last node from the list and returns it.
        """
        if not self.head:
            return None

        current_node = self.head
        new_tail = current_node

        while current_node.next:
            new_tail = current_node
            current_node = current_node.next

        new_tail.next = None
        self.tail = new_tail
        self.length -= 1

        if self.length == 0:
            self.head = None
            self.tail = None

        return current_node

    def shift(self) -> Optional[Node]:
        """
        Removes the first node from the list and returns it.
        """
        if not self.head:
            return None

        head_node = self.head
        self.head = head_node.next
        self.length -= 1

        if self.length == 0:
            self.tail = None

        return head_node

    def unshift(self, val: Any) -> 'LinkedList':
        """
        Adds a new node with the given value to the beginning of the list.
        """
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node

        self.length += 1
        return self

    def get(self, index: int) -> Optional[Node]:
        """
        Returns the node at the specified index.
        """
        if index < 0 or index >= self.length:
            return None

        current_node = self.head
        for _ in range(index):
            assert current_node is not None
            current_node = current_node.next

        return current_node

    def set(self, index: int, data: Any) -> Optional[Node]:
        """
        Updates the value of the node at the specified index.
        """
        node = self.get(index)
        if not node:
            return None

        node.val = data
        return node

    def insert(self, index: int, val: Any) -> Optional['LinkedList']:
        """
        Inserts a new node with the given value at the specified index.
        """
        if index < 0 or index > self.length:
            return None

        if index == 0:
            return self.unshift(val)

        if index == self.length:
            return self.push(val)

        new_node = Node(val)
        prev_node = self.get(index - 1)

        if not prev_node:
            return None

        new_node.next = prev_node.next
        prev_node.next = new_node
        self.length += 1

        return self

    def remove(self, index: int) -> Optional[Node]:
        """
        Removes and returns the node at the specified index.
        """
        if index < 0 or index >= self.length:
            return None

        if index == 0:
            return self.shift()

        if index == self.length - 1:
            return self.pop()

        prev_node = self.get(index - 1)
        if not prev_node or not prev_node.next:
            return None

        removed_node = prev_node.next
        prev_node.next = removed_node.next
        self.length -= 1

        return removed_node

    def to_array(self) -> list[Any]:
        """
        Converts the linked list to a Python list.
        """
        result = []
        current = self.head

        while current:
            result.append(current.val)
            current = current.next

        return result

    def print(self) -> None:
        """
        Prints the linked list as a Python list.
        """
        print(self.to_array())

    def reverse(self) -> 'LinkedList':
        """
        Reverses the linked list in place.
        """
        if not self.head or not self.head.next:
            return self

        current = self.head
        prev = None
        self.tail = self.head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        self.head = prev
        return self
